File: client/test/gestureDiscrimination.test.ts
import { describe, it, expect } from 'vitest';
import {
  isDistinctiveShape,
  straightness,
  normalizeGesture,
  MAX_ENROLLMENT_STRAIGHTNESS,
  type Point,
} from '../src/utils/gestureNormalize';
import {
  isGestureMatch,
  DEFAULT_MATCH_DISTANCE_THRESHOLD,
  CONFIRMATION_DISTANCE_THRESHOLD,
} from '../src/utils/gestureRecognizer';

function poly(...corners: [number, number][]): Point[] {
  const out: Point[] = [];
  for (let s = 0; s < corners.length - 1; s++) {
    const [x0, y0] = corners[s];
    const [x1, y1] = corners[s + 1];
    for (let i = 0; i <= 40; i++) {
      const t = i / 40;
      out.push({ x: x0 + (x1 - x0) * t, y: y0 + (y1 - y0) * t });
    }
  }
  return out;
}

const SHAPES: Record<string, Point[]> = {
  L: poly([40, 40], [40, 230], [230, 230]),
  check: poly([50, 140], [110, 210], [230, 60]),
  Z: poly([40, 40], [230, 40], [40, 230], [230, 230]),
  N: poly([40, 230], [40, 40], [230, 230], [230, 40]),
  triangle: poly([140, 40], [230, 220], [50, 220], [140, 40]),
  U: poly([50, 40], [50, 220], [220, 220], [220, 40]),
};

const LINES: Record<string, Point[]> = {
  horizontal: poly([40, 140], [240, 140]),
  vertical: poly([140, 40], [140, 240]),
  diagonal: poly([50, 50], [230, 230]),
};

/** A genuine redraw: different position, different size, hand wobble. */
function redraw(base: Point[], wobble: number, seed: number): Point[] {
  let state = seed;
  const rand = () => {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    return state / 0x7fffffff;
  };
  const dx = (rand() - 0.5) * 60;
  const dy = (rand() - 0.5) * 60;
  const scale = 1 + (rand() - 0.5) * 0.5;
  return base.map((p) => ({
    x: p.x * scale + dx + (rand() - 0.5) * 2 * wobble,
    y: p.y * scale + dy + (rand() - 0.5) * 2 * wobble,
  }));
}

describe('Gesture discrimination', () => {
  it('keeps the thresholds identical to the Android constants', () => {
    // Both platforms are tuned from the same sweep; drift between them means a gesture that
    // works on one device and not the other.
    expect(DEFAULT_MATCH_DISTANCE_THRESHOLD).toBe(14.0);
    expect(CONFIRMATION_DISTANCE_THRESHOLD).toBe(15.0);
    expect(MAX_ENROLLMENT_STRAIGHTNESS).toBe(0.9);
  });

  it('never lets one distinct shape unlock another', () => {
    const names = Object.keys(SHAPES);
    const accepted: string[] = [];

    for (const enrolled of names) {
      const template = normalizeGesture(SHAPES[enrolled]);
      expect(template).not.toBeNull();
      for (const attacker of names) {
        if (attacker === enrolled) continue;
        if (isGestureMatch(SHAPES[attacker], template!)) {
          accepted.push(`${attacker} unlocked ${enrolled}`);
        }
      }
    }

    expect(accepted).toEqual([]);
  });

  it('accepts genuine redraws even when untidy', () => {
    let attempts = 0;
    let rejected = 0;

    for (const base of Object.values(SHAPES)) {
      const template = normalizeGesture(base)!;
      for (let seed = 1; seed <= 15; seed++) {
        for (const wobble of [4, 9, 14]) {
          attempts++;
          if (!isGestureMatch(redraw(base, wobble, seed * 37 + wobble), template)) rejected++;
        }
      }
    }

    // Reveal asks for three strokes, so a high per-stroke reject rate compounds badly.
    expect((rejected / attempts) * 100).toBeLessThanOrEqual(5);
  });

  it('refuses straight lines at enrollment but allows real shapes', () => {
    for (const [name, line] of Object.entries(LINES)) {
      expect(straightness(line), name).toBeGreaterThan(MAX_ENROLLMENT_STRAIGHTNESS);
      expect(isDistinctiveShape(line), name).toBe(false);
    }

    for (const [name, shape] of Object.entries(SHAPES)) {
      expect(isDistinctiveShape(shape), name).toBe(true);
    }
  });
});

File: android/app/src/test/java/com/enctxt/GestureDiscriminationTest.kt
package com.enctxt

import com.enctxt.core.gesture.GestureNormalizer
import com.enctxt.core.gesture.GesturePoint
import com.enctxt.core.gesture.GestureRecognizer
import com.enctxt.core.gesture.GestureRepository
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

/**
 * The false-accept / false-reject matrix behind the recognizer's thresholds.
 *
 * The thresholds used to sit at 28/30, where roughly one in nine distinct-shape pairs matched
 * each other — a wrong gesture could unlock a message. These cases pin the separation that
 * justifies 14/15, so a future "users complain it's fussy, bump the threshold" change has to
 * confront the security cost rather than silently reintroduce it.
 */
class GestureDiscriminationTest {

    private fun poly(vararg corners: Pair<Float, Float>): List<GesturePoint> {
        val out = mutableListOf<GesturePoint>()
        for (segment in 0 until corners.size - 1) {
            val (x0, y0) = corners[segment]
            val (x1, y1) = corners[segment + 1]
            for (i in 0..40) {
                val t = i / 40f
                out.add(GesturePoint(x0 + (x1 - x0) * t, y0 + (y1 - y0) * t))
            }
        }
        return out
    }

    /** Distinct shapes a person might plausibly choose, all with at least one turn. */
    private val shapes: Map<String, List<GesturePoint>> = mapOf(
        "L" to poly(40f to 40f, 40f to 230f, 230f to 230f),
        "check" to poly(50f to 140f, 110f to 210f, 230f to 60f),
        "Z" to poly(40f to 40f, 230f to 40f, 40f to 230f, 230f to 230f),
        "N" to poly(40f to 230f, 40f to 40f, 230f to 230f, 230f to 40f),
        "triangle" to poly(140f to 40f, 230f to 220f, 50f to 220f, 140f to 40f),
        "U" to poly(50f to 40f, 50f to 220f, 220f to 220f, 220f to 40f),
        "circle" to (0..100).map {
            val a = it / 100f * 2f * Math.PI.toFloat()
            GesturePoint(140f + 90f * cos(a), 140f + 90f * sin(a))
        },
        "S" to (0..80).map {
            val t = it / 80f
            GesturePoint(40f + 200f * t, 140f + 80f * sin(t * 2f * Math.PI.toFloat()))
        }
    )

    /** A genuine redraw: different position, different size, hand wobble. */
    private fun redraw(base: List<GesturePoint>, wobble: Float, seed: Int): List<GesturePoint> {
        val rng = Random(seed)
        val dx = (rng.nextFloat() - 0.5f) * 60f
        val dy = (rng.nextFloat() - 0.5f) * 60f
        val scale = 1f + (rng.nextFloat() - 0.5f) * 0.5f
        return base.map {
            GesturePoint(
                it.x * scale + dx + (rng.nextFloat() - 0.5f) * 2f * wobble,
                it.y * scale + dy + (rng.nextFloat() - 0.5f) * 2f * wobble
            )
        }
    }

    private fun matches(drawn: List<GesturePoint>, enrolled: List<GesturePoint>): Boolean {
        val template = GestureNormalizer.normalize(enrolled) ?: return false
        return GestureRecognizer.isMatch(drawn, template)
    }

    @Test
    fun noDistinctShapeUnlocksAnother() {
        val names = shapes.keys.toList()
        val accepted = mutableListOf<String>()

        for (enrolled in names) {
            for (attacker in names) {
                if (enrolled == attacker) continue
                if (matches(shapes[attacker]!!, shapes[enrolled]!!)) {
                    accepted.add("$attacker unlocked $enrolled")
                }
            }
        }

        assertTrue("Wrong shapes must never unlock: $accepted", accepted.isEmpty())
    }

    @Test
    fun genuineRedrawsAreAcceptedEvenWhenUntidy() {
        var attempts = 0
        var rejected = 0

        for ((_, base) in shapes) {
            for (seed in 1..15) {
                for (wobble in listOf(4f, 9f, 14f)) {
                    attempts++
                    if (!matches(redraw(base, wobble, seed * 37 + wobble.toInt()), base)) rejected++
                }
            }
        }

        // A little strictness is the price of zero false accepts, but the flow must stay usable:
        // reveal asks for three strokes, so a high per-stroke reject rate compounds badly.
        val rate = rejected * 100.0 / attempts
        assertTrue("False-reject rate too high for a 3-stroke reveal: %.1f%%".format(rate), rate <= 5.0)
    }

    @Test
    fun straightLinesAreRefusedAtEnrollment() {
        val lines = mapOf(
            "horizontal" to poly(40f to 140f, 240f to 140f),
            "vertical" to poly(140f to 40f, 140f to 240f),
            "diagonal" to poly(50f to 50f, 230f to 230f),
            "reverse diagonal" to poly(230f to 50f, 50f to 230f)
        )

        lines.forEach { (name, line) ->
            assertFalse("$name should be refused as a gesture", GestureNormalizer.isDistinctiveShape(line))
        }

        // And every shape a person would actually pick still passes.
        shapes.forEach { (name, shape) ->
            assertTrue("$name should be allowed", GestureNormalizer.isDistinctiveShape(shape))
        }
    }

    @Test
    fun theRepositoryRefusesToPersistAStraightLine() {
        val storage = FakeGestureStorage()
        val repository = GestureRepository(storage)

        assertFalse(repository.saveSequence("user", listOf(poly(40f to 140f, 240f to 140f))))
        assertFalse("Nothing may be written", repository.isConfigured("user"))

        assertTrue(repository.saveSequence("user", listOf(shapes["L"]!!)))
        assertEquals(1, repository.sequenceLength("user"))
    }
}
